### PR TITLE
Flight: Support Webpack v5

### PR DESCRIPTION
**Pull request by samcooke98**
## Summary

I was experimenting with the new Server components. I noticed that this didn't work in webpack 5 due to the use of some deprecated APIs. 

Additionally, I noticed the regex was wrong (https://github.com/facebook/react/commit/c3f621dc646e6c21eec6231751f1ca3c0c8535c1), or at least it wouldn't include `tsx` files as server components.

I totally understand if you only wish to support webpack v4 while you experiment with Server-components. However, I thought this PR would be useful when you decide to support webpack v5 - This is why I'm opening it as a draft PR. 

## Test Plan

I added a webpack 5 alias. I wrote a test fixture in `ReactFlightWebpackPlugin-test.js`. This tests ensure that the webpack compilation: 
- Outputs the manifest, with the expected result
- Completes the compilation with no errors.

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -139,5 +139,8 @@
     "version-check": "node ./scripts/tasks/version-check.js",
     "merge-fork": "node ./scripts/merge-fork/merge-fork.js",
     "replace-fork": "node ./scripts/merge-fork/replace-fork.js"
+  },
+  "dependencies": {
+    "webpack5": "npm:webpack@5"
```
Comment (samcooke98) [inline package.json:144]: Oops, this should be a devDependency.

File: packages/react-server-dom-webpack/src/ReactFlightWebpackPlugin.js [hunk 2]
```diff
@@ -94,6 +95,51 @@ export default class ReactFlightWebpackPlugin {
...
+      };
+    });
+    const href = pathToFileURL(resource).href;
+
+    if (href !== undefined) {
+      this.pluginOutput[href] = moduleExports;
+    }
+  };
+
+  //
```
Comment (samcooke98) [inline packages/react-server-dom-webpack/src/ReactFlightWebpackPlugin.js:137]:
```suggestion
```

File: packages/react-server-dom-webpack/src/ReactFlightWebpackPlugin.js [hunk 3]
```diff
@@ -94,6 +95,51 @@ export default class ReactFlightWebpackPlugin {
...
+    {
+      getExports,
+      resource,
+      chunkIds,
+    }: {getExports: () => string[], resource: string, chunkIds: string[]},
+  ) => {
+    // TODO: Hook into deps instead of the target module.
+    // That way we know by the type of dep whether to include.
+    // It also resolves conflicts when the same module is in multiple chunks.
+    if (!/\.client\.(tsx|jsx|js|ts)$/.test(resource)) {
```
Comment (samcooke98) [inline packages/react-server-dom-webpack/src/ReactFlightWebpackPlugin.js:117]: From reading the RFC, I think the valid extensions should be any file that webpack accepts? EG: `resolve.extensions` 

In other words, from my undertanding of the RFCS: This code should be valid. 
```
// // Parent.server.js

import Component from "Component.client.js";
import implA from "implA.client.js";
import Picture from "MyCatPhoto.client.png";

export default function Parent() {
  return <Component avatar={Picture} impl={implA} />
}
```

(from: https://github.com/reactjs/rfcs/blob/235f27c12aa893efd2378ec3c4a9b0b221641861/text/0000-server-module-conventions.md#implicit-client-references

File: packages/react-server-dom-webpack/src/ReactFlightWebpackPlugin.js [hunk 4]
```diff
@@ -163,56 +209,87 @@ export default class ReactFlightWebpackPlugin {
...
+
+        // ProcessAssests hook is only supported by webpack v5.
+        if (this.webpackVersion !== 'wp5') {
+          return;
+        }
+
+        compilation.hooks.processAssets.tap(
+          {
+            name: PLUGIN_NAME,
+            // We derive the manifest from the existing assets.
```
Comment (samcooke98) [inline packages/react-server-dom-webpack/src/ReactFlightWebpackPlugin.js:221]:
```suggestion
```

I don't think this comment adds anything

**Review by samcooke98 — COMMENTED** [review]
This is a quick, non complete self review. I'll do a more indepth one before making this a proper PR, if there is appetite for this PR

**Review by sibelius — APPROVED** [review]

**Comment by gaearon** [comment]
Thanks for the early exploration! We merged https://github.com/facebook/react/pull/22739 which updates to Webpack 5 only. If you see something there that's wrong (and that your PR covers), please let me know!

The "automated test" part of this PR is interesting. We're going v5+ only, but I'm curious if you'd like to resubmit this part as a separate PR. Note that we wouldn't want to hardcode the output — but a snapshot test to prevent regressions seems useful.

**Comment by samcooke98** [comment]
> The "automated test" part of this PR is interesting. We're going v5+ only, but I'm curious if you'd like to resubmit this part as a separate PR. Note that we wouldn't want to hardcode the output — but a snapshot test to prevent regressions seems useful.


Happy to resubmit that in a separate PR, with a snapshot test instead. Will open one shortly, and close this one.

**Pull request closed without merging**